File: Dynamic_Programming/2786. Visit Array Positions to Maximize Score.cpp
/*
2786. Visit Array Positions to Maximize Score
*/

// TC:O(n) SC:O(2*n)
class Solution {
public:
    long long maxScore(vector<int>& nums, int x) {
        int n = nums.size();
        vector<vector<long long>> dp(n, vector<long long>(2, INT_MIN));
        if(nums[0] % 2 == 1){
            dp[0][1] = nums[0];
        }
        else {
            dp[0][0] = nums[0];
        }
        
        long long res = 0;
        for(int i = 1; i < n ; i++){
            //不取的狀態
            dp[i][0] = dp[i-1][0];
            dp[i][1] = dp[i-1][1];
            //取的話
            if(nums[i] % 2 == 1){
                dp[i][1] = max(dp[i][1], max(dp[i-1][1] + nums[i], dp[i-1][0] + nums[i] - x));
            }
            else {
                dp[i][0] = max(dp[i][0], max(dp[i-1][0] + nums[i], dp[i-1][1] + nums[i] - x));
            }
        }

        res = max(dp[n-1][0], dp[n-1][1]);
        return res;
    }
};

/*
此題說 你可以從起點走到後面任意點 可以累積分數 
但是如果這點跟你的上一點奇偶性不同 你就會失去x 分 
問最多能得到幾分
dp[i][j] : 走到i element時能得到的最大分 並且奇偶性是 % 2 == j

*/

File: Dynamic_Programming/2786. Visit Array Positions to Maximize Score.py
class Solution:
    def maxScore(self, nums: List[int], x: int) -> int:
        n = len(nums)
        # 空間優化
        odd = nums[0] if (nums[0] % 2) else -inf
        even = nums[0] if (nums[0] % 2) == 0 else -inf
        for i in range(1, n):
            if nums[i] % 2:
                odd = max(odd, even - x) + nums[i]
            else:
                even = max(even, odd - x) + nums[i]
        return max(odd, even)
        '''
        遞推
        f = [[-inf] * 2 for _ in range(n)]
        f[0][nums[0] % 2] = nums[0]
        for i in range(1, n):
            f[i][nums[i] % 2] = max(f[i-1][nums[i] % 2], f[i-1][1 - (nums[i] % 2)] - x) + nums[i]
            f[i][1 - nums[i] % 2] = f[i-1][1 - nums[i] % 2]
        return max(f[n-1][0], f[n-1][1])
        '''
        '''
        遞歸
        @cache
        def dfs(i: int, parity: int) -> int:
            if i == n:
                return 0
            ans = dfs(i + 1, parity)
            ans = max(ans, dfs(i + 1, nums[i] % 2) + nums[i]) if nums[i] % 2 == parity else max(ans, dfs(i + 1, nums[i] % 2) + nums[i] - x)
            return ans
        return nums[0] + dfs(1, nums[0] % 2)
        '''
